wandb/sdk/launch/runner/kubernetes_monitor.py

```diff
@@ -0,0 +1,297 @@
+    # Starting states.
+    "created": "starting",
+    "pending": "starting",
+    # Running states.
+    "running": "running",
+    "completing": "running",
+    # Finished states.
+    "succeeded": "finished",
+    "completed": "finished",
+    # Failed states.
+    "failed": "failed",
+    "aborted": "failed",
+    "timeout": "failed",
+    "terminated": "failed",
+    # Stopping states.
+    "terminating": "stopping",
+}
+
+
+_logger = logging.getLogger(__name__)
+
+
+class SafeWatch:
+    """Wrapper for the kubernetes watch class that can recover in more situations."""
+
+    def __init__(self, watcher: "watch.Watch") -> None:
+        """Initialize the SafeWatch."""
+        self._watcher = watcher
+        self._stopped = False
+
+    def stream(self, func: Any, *args: Any, **kwargs: Any) -> Any:
+        """Stream the watcher."""
+        while True:
+            try:
+                for event in self._watcher.stream(
+                    func, *args, **kwargs, timeout_seconds=15
+                ):
+                    # Save the resource version so that we can resume the stream
+                    # if it breaks.
+                    kwargs["resource_version"] = event.get(
+                        "object"
+                    ).metadata.resource_version
+                    yield event
+                # If stream ends after stop just break
+                if self._stopped:
+                    break
+            except urllib3.exceptions.ProtocolError as e:
+                wandb.termwarn(f"Broken event stream: {e}")
+            except ApiException as e:
+                if e.status == 410:
+                    # If resource version is too old we need to start over.
+                    del kwargs["resource_version"]
+            except Exception as E:
+                wandb.termerror(f"Unknown exception in event stream: {E}")
+
+    def stop(self) -> None:
+        """Stop the watcher."""
+        self._watcher.stop()
+        self._stopped = True
+
+
+def _is_preempted(status: "V1PodStatus") -> bool:
+    """Check if this pod has been preempted."""
+    if hasattr(status, "conditions") and status.conditions is not None:
+        for condition in status.conditions:
+            if condition.type == "DisruptionTarget" and condition.reason in [
+                "EvictionByEvictionAPI",
+                "PreemptionByScheduler",
+                "TerminationByKubelet",
+            ]:
+                return True
+    return False
+
+
+def _is_container_creating(status: "V1PodStatus") -> bool:
+    """Check if this pod has started creating containers."""
+    for container_status in status.container_statuses or []:
+        if (
+            container_status.state
+            and container_status.state.waiting
+            and container_status.state.waiting.reason == "ContainerCreating"
+        ):
+            return True
+    return False
+
+
+def _state_from_conditions(conditions: List[Dict[str, Any]]) -> Optional[str]:
```

Comment:
I think there are some incorrect assumptions here. Each entry in `conditions` is independent of any other entry.  Multiple conditions can be true at the same time, but conditions are unique by `type` for a given resource.  As implemented, this will give you the latest state w/r/t transition time, but it's unclear what guarantees we have around that.  For example, look at this `status` block for a completed job:

```
        "conditions": [
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2023-09-12T18:25:30Z",
                "reason": "PodCompleted",
                "status": "True",
                "type": "Initialized"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2023-09-12T18:25:30Z",
                "reason": "PodCompleted",
                "status": "False",
                "type": "Ready"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2023-09-12T18:25:30Z",
                "reason": "PodCompleted",
                "status": "False",
                "type": "ContainersReady"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2023-09-12T18:25:30Z",
                "status": "True",
                "type": "PodScheduled"
            }
        ],
```

every single one has an identical timestamp, so the sort is arbitrary, and I'd expect this to return either `PodScheduled` or `PodCompleted`, depending on the order it came in (did not find any guarantees w/r/t this in the docs).  Instead, this function should probably return a `List[str]` for all conditions where `status` is `True`.  Within that list is a subset of conditions we actually care about.  So the consuming code should check for membership not equality.

Reply:
Yep you're 100% right.

Reply:
I've updated this to instead look at all true conditions and return a status from one as soon as we find a match. This does make the assumption that the list of true conditions doesn't contain 2 or more types that correspond to different statuses

For jobset the conditions are only used to mark the final state of a job, e.g.
```
conditions:
    - lastTransitionTime: "2023-09-13T02:16:15Z"
      message: jobset completed successfully
      reason: AllJobsCompleted
      status: "True"
      type: Completed
```

is the only condition we ever see for a successful jobset run. For a kubeflow job the set of conditions will look like:
```
conditions:
  - lastTransitionTime: "2023-08-30T21:32:45Z"
    lastUpdateTime: "2023-08-30T21:32:45Z"
    message: PyTorchJob xe02v24m-pytorch-job is created.
    reason: PyTorchJobCreated
    status: "True"
    type: Created
  - lastTransitionTime: "2023-08-30T21:32:47Z"
    lastUpdateTime: "2023-08-30T21:32:47Z"
    message: PyTorchJob xe02v24m-pytorch-job is running.
    reason: PyTorchJobRunning
    status: "False"
    type: Running
  - lastTransitionTime: "2023-08-30T21:32:58Z"
    lastUpdateTime: "2023-08-30T21:32:58Z"
    message: PyTorchJob xe02v24m-pytorch-job is successfully completed.
    reason: PyTorchJobSucceeded
    status: "True"
    type: Succeeded
```

Dammit I actually see here that we can have created and succeeded true at the same time for kubeflow pytorchjobs so I think I am going to bring back sorting based on timestamp. Or just get the whole set of states implied by the statuses and resolve their precedence with common sense, like running or succeeded takes precedence over created. Terminating takes precedence over running etc.

Reply:
In the spirit of this PR I should definitely be examining all conditions rather than just the most recently updated one since watches can need to be entirely restarted.